crates/red_knot_ide/src/lib.rs

```diff
@@ -1,3 +1,222 @@
+    ///
+    /// The `focus_range` must be fully covered by `full_range`.
+    focus_range: TextRange,
+
+    /// The range covering the entire target.
+    full_range: TextRange,
+}
+
+impl NavigationTarget {
+    pub fn file(&self) -> File {
+        self.file
+    }
+
+    pub fn focus_range(&self) -> TextRange {
+        self.focus_range
+    }
+
+    pub fn full_range(&self) -> TextRange {
+        self.full_range
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct NavigationTargets(smallvec::SmallVec<[NavigationTarget; 1]>);
+
+impl NavigationTargets {
+    fn single(target: NavigationTarget) -> Self {
+        Self(smallvec::smallvec![target])
+    }
+
+    fn empty() -> Self {
+        Self(smallvec::SmallVec::new())
+    }
+
+    fn iter(&self) -> std::slice::Iter<'_, NavigationTarget> {
+        self.0.iter()
+    }
+
+    #[cfg(test)]
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl IntoIterator for NavigationTargets {
+    type Item = NavigationTarget;
+    type IntoIter = smallvec::IntoIter<[NavigationTarget; 1]>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a NavigationTargets {
+    type Item = &'a NavigationTarget;
+    type IntoIter = std::slice::Iter<'a, NavigationTarget>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+impl FromIterator<NavigationTarget> for NavigationTargets {
+    fn from_iter<T: IntoIterator<Item = NavigationTarget>>(iter: T) -> Self {
+        Self(iter.into_iter().collect())
+    }
+}
+
+pub trait HasNavigationTargets {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets;
+}
+
+impl HasNavigationTargets for Type<'_> {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets {
+        match self {
+            Type::FunctionLiteral(function) => function.navigation_targets(db),
+            Type::ModuleLiteral(module) => module.navigation_targets(db),
+            Type::Union(union) => union
+                .iter(db.upcast())
+                .flat_map(|target| target.navigation_targets(db))
+                .collect(),
+            Type::ClassLiteral(class) => class.navigation_targets(db),
+            Type::Instance(instance) => instance.navigation_targets(db),
+            Type::KnownInstance(instance) => instance.navigation_targets(db),
+            Type::StringLiteral(_)
+            | Type::AlwaysTruthy
+            | Type::AlwaysFalsy
+            | Type::IntLiteral(_)
+            | Type::BooleanLiteral(_)
+            | Type::LiteralString
+            | Type::BytesLiteral(_)
+            | Type::SliceLiteral(_) => self.to_meta_type(db.upcast()).navigation_targets(db),
+
+            Type::Dynamic(_)
+            | Type::SubclassOf(_)
+            | Type::Never
+            | Type::Callable(_)
+            | Type::Intersection(_)
```

Comment:
Why not handle this the same as union and collect all navigation targets from elements?

(I think we can ignore negative elements.)

Reply:
The difference is that the value isn't the type of any of the elements in the intersection. Instead, it's the type that intersects with all elements. So I think it's just wrong? I can try and see what typescript does here

Reply:
It's the type of _all_ the elements of the intersection; if anything it seems even more correct here than in the union case? In the union case, the actual value only needs to be one of the union elements, the other ones are all "wrong".

Reply:
TypeScript doesn't support *goto type definition* on intersections, which matches my understanding unless Red Knot's intersection types work differently.

The LSP shows one location for each union variant, which very nicely matches how unions work: The value is one of the types shown (see screenshot above). 

It would be very surprising if we showed intersections exactly the same as unions because users would then be unable to distinguish them from unions. I'll leave this as is (minus filtering out `Unknown`) because I'm not convinced that presenting them the same as unions is the right solution and there isn't an obvious alternative.